components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/flows/finality/ConsensualFinalityFlow.kt

```diff
@@ -47,146 +48,109 @@ class ConsensualFinalityFlow(
             }
-        }
+            session to signatures
+        }.toMap()
 
-        return signedByParticipantsTransaction
+        return transaction to signaturesReceivedFromSessions
     }
 
-    private fun requireCorrectReceivedSigningKeys(
-        signatures: List<DigitalSignatureAndMetadata>,
-        missingSignatories: Set<PublicKey>,
-        ledgerKeys: List<PublicKey>,
-        session: FlowSession
+    private fun verifyAllReceivedSignatures(
+        transaction: ConsensualSignedTransactionInternal,
+        signaturesReceivedFromSessions: Map<FlowSession, List<DigitalSignatureAndMetadata>>
     ) {
-        val receivedSigningKeys = signatures.map { it.by }
-        val expectedSigningKeys = missingSignatories.intersect(ledgerKeys.toSet())
-        if (receivedSigningKeys.toSet() != expectedSigningKeys) {
-            throw CordaRuntimeException(
-                "A session with ${session.counterparty} did not return the signatures with the expected keys. " +
-                        "Expected: $expectedSigningKeys But received: $receivedSigningKeys"
-            )
+        log.debug { "Verifying all signatures for transaction $transactionId." }
+
+        try {
+            transaction.verifySignatures()
+        } catch (e: TransactionMissingSignaturesException) {
+            val counterpartiesToSignatoriesMessages = signaturesReceivedFromSessions.map { (session, signatures) ->
+                "${session.counterparty} provided ${signatures.size} signature(s) to satisfy the signatories (encoded) " +
+                        signatures.map { it.by.encoded }
+            }
+            val counterpartiesToSignatoriesMessage = if (counterpartiesToSignatoriesMessages.isNotEmpty()) {
+                "\n${counterpartiesToSignatoriesMessages.joinToString(separator = "\n")}"
+            } else {
+                "[]"
+            }
+            val message = "Transaction $transactionId is missing signatures for signatories (encoded) " +
+                    "${e.missingSignatories.map { it.encoded }}. The following counterparties provided signatures while finalizing " +
+                    "the transaction: $counterpartiesToSignatoriesMessage"
+            log.warn(message)
+            throw TransactionMissingSignaturesException(transactionId, e.missingSignatories, message)
         }
     }
 
     @Suspendable
-    private fun persistFullySignedTransaction(fullySignedTransaction: ConsensualSignedTransactionInternal) {
-        persistenceService.persist(fullySignedTransaction, TransactionStatus.VERIFIED)
-        log.debug { "Recorded signed transaction ${signedTransaction.id}" }
+    private fun persistTransactionWithCounterpartySignatures(transaction: ConsensualSignedTransactionInternal) {
+        persistenceService.persist(transaction, TransactionStatus.VERIFIED)
+        log.debug { "Recorded transaction $transactionId" }
```

Comment:
I would make this log more specific.
```suggestion
        log.debug { "Recorded transaction $transactionId with all signatures." }
```
?